Add versioning to MIOpen config and cache paths

JehandadKhan opened the pull request:
To reduce the chance of collisions between schema as a result of version change, this PR versions the MIOpen config and cache directories.

atamazov commented:
I see several aspects to discuss. _What_ we are going to change and _why_, plus _how_ we are doing that.

First about "_how_".

MIOpen versioning is already used for find-db. files I highly recommend employing the similar approach, i.e. encoding the version information into the filename. Please see https://github.com/AMDComputeLibraries/MLOpen/issues/1139 for some discussion. Implementation generates parts of the filename ("suffixes") and then using these suffixes as appropriate.

To avoid duplication, I would recommend defining the variable for version info to be encoded into database paths. I would propose something like
```
set(MIOPEN_VERSION_SUFFIX "${MIOpen_VERSION_MAJOR}_${MIOpen_VERSION_MINOR}_${MIOpen_VERSION_PATCH}" CACHE PATH "Allows different versions of the library to use different database files")
```
This can be reused in definition of `MIOPEN_USER_DB_SUFFIX`.

About "_what_ we would like to do and _why_" -- this topic worth ticket, I think. I would better dedicate a ticket to it. However we can discuss right here, if you would like to. Please let me know and I will proceed.

JehandadKhan commented:
@atamazov we can discuss it here.

atamazov commented:
@JehandadKhan okay, thanks.

I am assuming that the purpose is to support several installations of MIOpen (built with different installations of ROCm) on the same machine, right?

The obvious solution is including ROCm version into the db pathnames. What I am missing?

What is the purpose of "tweak" in the db pathname?

I think that need to change the User find-db pathname as well.

Is it so that OCL and HIP backends share the same binary cache and perf-db?

Do we need to support two installations of MIOpen built with different backends (OpenCL and HIP) on the same machine? (single ROCm installation)

daniellowell commented:
This change is required by this blocking ticket:
http://ontrack-internal.amd.com/browse/SWDEV-220965

The only change we require is to separate the user files based on the ROCm version components used.

> What is the purpose of "tweak" in the db pathname?
The purpose is to associate a version of MIOpen that is build with the components of a specific ROCm release's tools.

>Do we need to support two installations of MIOpen built with different backends (OpenCL and HIP) on the same machine? (single ROCm installation)

No.

daniellowell commented on CMakeLists.txt:
```diff
@@ -262,8 +262,8 @@ if(BUILD_DEV)
     set(MIOPEN_SYSTEM_DB_PATH "${CMAKE_INSTALL_PREFIX}/${DATA_INSTALL_DIR}/db" CACHE PATH "Default path of system db files")
-    set(MIOPEN_USER_DB_PATH "~/.config/miopen/" CACHE STRING "Default path of user db files")
-    set(MIOPEN_CACHE_DIR "~/.cache/miopen/" CACHE STRING "")
+    set(MIOPEN_USER_DB_PATH "~/.config/miopen.${MIOpen_VERSION_MAJOR}_${MIOpen_VERSION_MINOR}_${MIOpen_VERSION_PATCH}_${MIOpen_VERSION_TWEAK}/" CACHE STRING "Default path of user db files")
```
This will result in `~/.cache/miopen.M.m.p.t/M.m.p/miopen.udb` I think. Example: `~/.cache/miopen.2.4.0.439824/2.4.0/miopen.udb`.
Don't we instead want:
`~/.cache/miopen/M.m.p.t/miopen.udb`
For example:
`~/.cache/miopen/2.4.0.439824/miopen.udb`

atamazov replied:
This will also move user find-db files from `~/.config/` to "versioned" directory, but these files already contains version in db filenames: `~/.config/miopen/2.4.0.439824/gfx906_60.HIP.2_4_0.ufdb.txt`, so we are doing the same job twice.

The idea for perf-db and find-db was to use paths to denote base directory, and encode whatever is necessary (for differentiating databases) into filenames.

The filesystem-based binary cache uses subdirectories, so the "separating" information is encoded into the name of the topmost directory. Now we do not use subdirs, but the directory remains the same.

(Current implementation is messed up :face_with_head_bandage:  Perhaps it's time to make it uniform, but I still do not have clear idea.)

pfultz2 commented on CMakeLists.txt:
```diff
@@ -262,8 +262,8 @@ if(BUILD_DEV)
-    set(MIOPEN_USER_DB_PATH "~/.config/miopen/" CACHE STRING "Default path of user db files")
-    set(MIOPEN_CACHE_DIR "~/.cache/miopen/" CACHE STRING "")
+    set(MIOPEN_USER_DB_PATH "~/.config/miopen.${MIOpen_VERSION_MAJOR}_${MIOpen_VERSION_MINOR}_${MIOpen_VERSION_PATCH}_${MIOpen_VERSION_TWEAK}/" CACHE STRING "Default path of user db files")
+    set(MIOPEN_CACHE_DIR "~/.cache/miopen.${MIOpen_VERSION_MAJOR}_${MIOpen_VERSION_MINOR}_${MIOpen_VERSION_PATCH}_${MIOpen_VERSION_TWEAK}/" CACHE STRING "")
```
The version for the cache happens [here](https://github.com/ROCmSoftwarePlatform/MIOpen/blob/develop/src/binary_cache.cpp#L63) not in cmake. That is where the tweak should be added.

pfultz2 commented on CMakeLists.txt:
```diff
@@ -262,8 +262,8 @@ if(BUILD_DEV)
     set(MIOPEN_SYSTEM_DB_PATH "${CMAKE_INSTALL_PREFIX}/${DATA_INSTALL_DIR}/db" CACHE PATH "Default path of system db files")
-    set(MIOPEN_USER_DB_PATH "~/.config/miopen/" CACHE STRING "Default path of user db files")
-    set(MIOPEN_CACHE_DIR "~/.cache/miopen/" CACHE STRING "")
+    set(MIOPEN_USER_DB_PATH "~/.config/miopen.${MIOpen_VERSION_MAJOR}_${MIOpen_VERSION_MINOR}_${MIOpen_VERSION_PATCH}_${MIOpen_VERSION_TWEAK}/" CACHE STRING "Default path of user db files")
```
I dont think we should version the db here. We should only version the finddb(since its fast to compute). We shouldn't version the perfdb. No one wants to spend a week to re-tune because only the rocm version changed. 

Users of multi-rocm will most likely not use perfdb(and we make it clear not use it in reproducible multi-install environments) since it requires extra special steps to create. The finddb, it makes sense to version as that will get updated by when users call `Find`.

atamazov commented:
@daniellowell Thanks for feedback.

> This change is required by this blocking ticket: http://ontrack-internal.amd.com/browse/SWDEV-220965

[Clarification] This is P2 "must solve", not a P1 blocker.

> The only change we require is to separate the user files based on the ROCm version components used.

[Clarification] Separation needs to be done for the binary cache and for find-db. WRT perf-db I agree with @pfultz2 that this is not needed.

>> What is the purpose of "tweak" in the db pathname?

> The purpose is to associate a version of MIOpen that is build with the components of a specific ROCm release's tools.

But TWEAK does not depend on ROCm release's tools.

Perhaps I am too dumb, but let me repeat my previous question: ___The obvious solution is including ROCm version into db pathnames.___ What I am missing?

daniellowell commented:
It is a P2 for now, but it is a requirement to get this all nailed down so I am treating it like a P1.

>WRT perf-db I agree with @pfultz2 that this is not needed.

I don't see how that is going to work. The data base is continuously updated in functionality. We are adding 3D convolutions which means the ~/.config/miopen/miopen.udb will become obsolete. If that isn't versioned then that's a problem.

>But TWEAK does not depend on ROCm release's tools.

True, but it does have a build number. That should be suffienct.

Versioning the individual files seems less tidy than versioning their directories. I mean we can do both if that makes you happy. It seems confusing if there are 5 version of the same file under 2.4.0 if we don't change the version of MIOpen over 5 releases (for example).

pfultz2 commented:
> I don't see how that is going to work. The data base is continuously updated in functionality. We are adding 3D convolutions which means the ~/.config/miopen/miopen.udb will become obsolete. If that isn't versioned then that's a problem.

What do you mean obsolete? Is the schema changing? Isn't 3D convolutions just a different key lookup?

We probably do need to version perfdb but it should based on if we make backwards compatible changes instead of using the rocm version.

pfultz2 commented:
> But TWEAK does not depend on ROCm release's tools.

It does. In rocm 3.0, the version was `2.2.0.7617-rocm-rel-3.0-6-948938d` and in 3.1 it was `2.2.1.7633-rocm-rel-3.1-35-9218683`. That last part of the version is the tweak. It gets set to a commit number(ie 7617), the build id(ie rocm-rel-3.0-6) and the commit hash(ie 948938d).

daniellowell commented:
>What do you mean obsolete? Is the schema changing? Isn't 3D convolutions just a different key lookup?

Schema yeah, for 3D convolutions. @JehandadKhan I think that should it should just be insertion warning if an new 3D record is going to be inserted into an, so that we maintain backward compatibility if users want to keep their old user performance database.

pfultz2 commented:
> Schema yeah, for 3D convolutions.

Why is it a schema change?

JehandadKhan commented:
> > Schema yeah, for 3D convolutions.
> 
> Why is it a schema change?

Because we added new fields, unlike a text file we do not concatenate string to form a key, the different parameters are fields in a table which get a SELECT query on them. Since 3D convolutions require new fields, therefore the schema changed.

JehandadKhan commented:
> > Schema yeah, for 3D convolutions.
> 
> Why is it a schema change?

We can do that programmatically by being more smart about creating the database. Currently we don't do that since it requires the database being locked for longer during creation.

pfultz2 commented:
> Since 3D convolutions require new fields, therefore the schema changed.

But shouldn't that be in a different table?

JehandadKhan commented:
> > Since 3D convolutions require new fields, therefore the schema changed.
> 
> But shouldn't that be in a different table?

Adding columns to a table is also a schema change.

JehandadKhan commented:
@daniellowell , @atamazov , @pfultz2 do we have a consensus on what we want to do ?

pfultz2 commented:
I think we should version perfdb but only change the version when there are non-backwards compatible changes(such as schema changes). It shouldn't be based on the rocm version.

daniellowell commented:
>I think we should version perfdb but only change the version when there are non-backwards compatible changes(such as schema changes). It shouldn't be based on the rocm version.

That is what I think as well. To be specific we are talking about the user perf db versioning.

atamazov commented:
@pfultz2 
> It does. In rocm 3.0, the version was `2.2.0.7617-rocm-rel-3.0-6-948938d` and in 3.1 it was `2.2.1.7633-rocm-rel-3.1-35-9218683`. That last part of the version is the tweak. It gets set to a commit number(ie 7617), the build id(ie rocm-rel-3.0-6) and the commit hash(ie 948938d).

What I'am seeing on my ROCm 3.0 machine differs a bit. It is `2.4.0.7760-a02171c92`, where `a02171c92` is a MIOpen commit hash. However `7760` looks like commit number and should be sufficient for our purposes.

It seems like we can use tweak to detect ROCm version and build number info, so I will eventually try to use it for "compiler version detection" instead of checks at run time.

Thanks for the explanation.

atamazov commented:
@JehandadKhan 
> do we have a consensus on what we want to do ?

I think so, yes.

JehandadKhan commented:
@atamazov @pfultz2 I have updated the code in light of the discussion above, a few salients:

* SQLite schema has its own version now, which is set in the CMake file.
* I have removed the subdirectories for the MIOpen version. This appeared to be redundant now.
* I have applied the changes to the perf db and the binary cache.

pfultz2 commented:
> It seems like we can use tweak to detect ROCm version and build number info, so I will eventually try to use it for "compiler version detection" instead of checks at run time.

You should not use this for compiler version detected. The compiler version could be unrelated to the rocm version. We install newer compiler version on older rocm all the time.

Instead, you should use feature detection instead(ie try to compile a test program to see if it will pass). This can be done at cmake time or at runtime. I believe this was done before.

pfultz2 commented on src/binary_cache.cpp:
```diff
@@ -60,11 +60,7 @@ boost::filesystem::path ComputeUserCachePath()
-                          std::to_string(MIOPEN_VERSION_PATCH);
-
-    auto p = boost::filesystem::path{miopen::ExpandUser(cache_dir)} / version;
+    auto p = boost::filesystem::path{miopen::ExpandUser(cache_dir)};
```
The binary cache should be versioned.

JehandadKhan replied:
Yeah, please see below

JehandadKhan replied:
Fixed

pfultz2 commented:
Also, the versioning for finddb needs to include the tweak if it isn't already.

pfultz2 commented on src/binary_cache.cpp:
```diff
@@ -98,8 +94,9 @@ KDb GetDb(const std::string& device, size_t num_cu)
-    boost::filesystem::path sys_path = sys_dir / (Handle::GetDbBasename(device, num_cu) + ".kdb");
+        user_dir / (Handle::GetDbBasename(device, num_cu) + "_" MIOPEN_SQLITE_SCHEMA_VER ".ukdb");
+    boost::filesystem::path sys_path =
+        sys_dir / (Handle::GetDbBasename(device, num_cu) + "_" MIOPEN_SQLITE_SCHEMA_VER ".kdb");
```
The schema version is for the perfdb not the binary cache.

JehandadKhan replied:
I think a separate version number for each database is excessive, particularly since these may be merged into one

pfultz2 replied:
What do you mean?

JehandadKhan replied:
While, currently we have a separate database for perf db, kernel cache and (in the future) find dd. They are essentially tables which may be part of the same `db` file. Thus may be merged.

pfultz2 replied:
I see what you meant.

There is only one schema version for perfdb. The finddb and binary cache will use the miopen version.

Finddb already writes the miopen version in the name so it just needs to include the tweak version.

The binary cache already writes a separate subdirectory for each version so it just needs to be extended to include the tweak version.

JehandadKhan replied:
Fixed

pfultz2 commented on CMakeLists.txt:
```diff
@@ -51,6 +51,7 @@ include(ROCMInstallSymlinks)
 include(ROCMCreatePackage)
 include(CheckCXXCompilerFlag)
 
+set(MIOPEN_SQLITE_SCHEMA_VER "1.0.0" CACHE STIRNG "Version of the sqlite schema")
```
Maybe this should be renamed to `MIOPEN_PERFDB_SCHEMA_VER` since this is the schema version for perfdb. 

And why does this need to be a cmake variable? I dont see any reason the user will override this when invoking cmake. Why not make the version a const variable in the source?

JehandadKhan replied:
I agree, the user does not need to over ride it. I would prefer to keep it in cmake since its a central location and makes management easier.

JehandadKhan replied:
I can put it in config.h, would that be ok ?

JehandadKhan replied:
Fixed

JehandadKhan commented:
@daniellowell @pfultz2 @atamazov All reviewer comments addressed

pfultz2 commented on CMakeLists.txt:
```diff
@@ -265,7 +265,7 @@ else()
     set(MIOPEN_CACHE_DIR "~/.cache/miopen/" CACHE STRING "")
 endif()
-set(MIOPEN_USER_DB_SUFFIX "${MIOPEN_BACKEND}.${MIOpen_VERSION_MAJOR}_${MIOpen_VERSION_MINOR}_${MIOpen_VERSION_PATCH}" CACHE PATH "Filename suffix for the user find-db files")
+set(MIOPEN_USER_DB_SUFFIX "${MIOPEN_BACKEND}.${MIOpen_VERSION_MAJOR}_${MIOpen_VERSION_MINOR}_${MIOpen_VERSION_PATCH}_${MIOpen_VERSION_TWEAK}" CACHE PATH "Filename suffix for the user find-db files")
```
We dont want the tweak version for the perfdb just for the finddb.

JehandadKhan replied:
This variable is only used in find db, I removed the use in the text based perf db as well

atamazov commented on src/include/miopen/execution_context.hpp:
```diff
@@ -58,6 +58,7 @@ class rocm_meta_version
 
 namespace miopen {
 
+const char* const MIOPEN_PERFDB_SCHEMA_VER = "1.0.0";
```
This defines the global pointer and C-string in each module where this header is included. Perhaps modern C++ "optimizations" can optimize this out, but this... I do not believe that we want this.

pfultz2 replied:
It can be in the header but it should be moved to the class level so it will point to the same address across all TUs.

JehandadKhan replied:
Moved to a  static member

atamazov commented on src/include/miopen/execution_context.hpp:
```diff
@@ -100,12 +101,11 @@ struct ExecutionContext
              + GetStream().GetDbBasename()
              + "."
-             + GetUserDbSuffix()
+             + std::string{MIOPEN_PERFDB_SCHEMA_VER}
```
Text-based perf-db does not need this, could you please revert.

atamazov replied:
oops, `GetUserDbSuffix()` is removed, looks like that this change was made by accident.

JehandadKhan replied:
No, this was requested by @pfultz2.

atamazov replied:
@JehandadKhan @pfultz2 Hmm. This specific change is wrong because is removes backend name and miopen version from user-find-db name and adds SQL db schema version that is not relevant.

There is RFC ticket related to find-db naming, https://github.com/AMDComputeLibraries/MLOpen/issues/1139. Let's discuss & agree there prior making changes.

JehandadKhan replied:
this line is for the text version of user perf db, which is not backend dependent ?

atamazov replied:
@JehandadKhan @pfultz2 Oops, sorry, I've mixed up different databases ;( Brand new comment:

This change is incorrect because is removes miopen version from the text-based user-perf-db. Also, SQL db schema version is not relevant here.

I do not have specific ticket for the _perf-db naming_, so let's discuss right here.
- Library version is necessary; this is obvious.
- Text-based databases do not change format (or backward compatible at least, -- this is a requirement).

The change also removes backend name, but this question is more complex.
- Related discussion/explanation can be found around [this comment](https://github.com/AMDComputeLibraries/MLOpen/issues/1419#issuecomment-545114745).
- In short, we can live with backend-agnostic system perf-db, but we still need backend-specific user layer ensures correct work in corner cases.
- These corner cases may happen when the user wants _two installations with different backends on the same system_. The one may think that such a case is purely theoretical, however developers can switch backends easily (and I do that often).
- I think that _if we do not want to support such setups, then backend name can be removed from the user-perf-db name._

JehandadKhan replied:
I have reverted the change, since its not relevant to SQLite Perf Db

atamazov replied:
Resolved. Please do not click `[Resolve conversation]` to keep this visible for a while.

atamazov commented on src/include/miopen/execution_context.hpp:
```diff
@@ -100,12 +101,11 @@ struct ExecutionContext
         return GetUserDbPath()
 #if MIOPEN_ENABLE_SQLITE
-             + "/miopen.udb";
+             + "miopen_" + std::string{MIOPEN_PERFDB_SCHEMA_VER} + ".udb";
```
[tip] You can simply use `string{"1.0.0"}` right here.

JehandadKhan replied:
I don't think burying version strings in the code is a good idea.

aserio commented:
@JehandadKhan, would you be able to get this change in this week?

JehandadKhan commented:
@aserio Yes, if the CI passes.

aserio commented:
Great! We have promised this feature for the next release.
